### PR TITLE
[ENH] Deep Clustering module basis with an auto encoder FCN based example

### PR DESCRIPTION
**Pull request by hadifawaz1999**
This is just a template now, i want to get ideas on how the module will be like, for now i put a up the base class and will see what to add later on in it.

### CODE REVIEW

**Comment by hadifawaz1999** [comment]
Added an example of a deep clusterer which is an auto-encoder based on the FCN network.
By doing this i added a new network which AE-FCN (Auto-Encoder FCN) and the deep clusterer.

The auto-encoder will be trained to reconstruct the input normally and the clustering will be then evaluated in its latent space (between encoder and decoder).

The clustering evaluation uses kmeans in the latent space.

File: aeon/clustering/deep_learning/ae_fcn.py [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+from aeon.networks import AEFCNNetwork
+from aeon.utils.validation._dependencies import _check_dl_dependencies
+
+
+class AEFCNClusterer(BaseDeepClusterer):
+    """Fully Connected Neural Network (FCN), as described in [1].
+
+    Parameters
+    ----------
+    n_clusters: int, default=None
```
Comment (TonyBagnall) [inline aeon/clustering/deep_learning/ae_fcn.py:24]: soz for being a pedant, but you need a space between variable and colon
n_clusters : int, default=None

File: aeon/clustering/deep_learning/ae_fcn.py [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+
+from sklearn.utils import check_random_state
+
+from aeon.clustering.deep_learning.base import BaseDeepClusterer
+from aeon.networks import AEFCNNetwork
+from aeon.utils.validation._dependencies import _check_dl_dependencies
+
+
+class AEFCNClusterer(BaseDeepClusterer):
+    """Fully Connected Neural Network (FCN), as described in [1].
```
Comment (TonyBagnall) [inline aeon/clustering/deep_learning/ae_fcn.py:20]: as described in [1]_
will link it in the API

File: aeon/clustering/deep_learning/ae_fcn.py [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+
+    Examples
+    --------
+    >>> from aeon.clustering.deep_learning import AEFCNClusterer
+    >>> from aeon.datasets import load_unit_test
+    >>> X_train, y_train = load_unit_test(split="train", return_X_y=True)
+    >>> X_test, y_test = load_unit_test(split="test", return_X_y=True)
+    >>> aefcn = AEFCNClusterer(n_clusters=2,n_epochs=20,batch_size=4)  # doctest: +SKIP
+    >>> aefcn.fit(X_train)  # doctest: +SKIP
+    FCNClassifier(...)
```
Comment (TonyBagnall) [inline aeon/clustering/deep_learning/ae_fcn.py:105]: should be 
AEFCNClusterer

File: aeon/clustering/deep_learning/ae_fcn.py [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+            padding=self.padding,
+            activation=self.activation,
+            use_bias=self.use_bias,
+            random_state=self.random_state,
+        )
+
+    def build_model(self, input_shape, **kwargs):
+        """Construct a compiled, un-trained, keras model that is ready for training.
+
+        In aeon, time series are stored in numpy arrays of shape (d,m), where d
```
Comment (TonyBagnall) [inline aeon/clustering/deep_learning/ae_fcn.py:195]: could you reword this to (n_channels, n_timepoints)?

File: aeon/clustering/deep_learning/ae_fcn.py [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+        is the number of dimensions, m is the series length. Keras/tensorflow assume
+        data is in shape (m,d). This method also assumes (m,d). Transpose should
+        happen in fit.
+
+        Parameters
+        ----------
+        input_shape : tuple
+            The shape of the data fed into the input layer, should be (m,d).
```
Comment (TonyBagnall) [inline aeon/clustering/deep_learning/ae_fcn.py:203]: same here

File: aeon/clustering/deep_learning/ae_fcn.py [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+        model.compile(optimizer=self.optimizer_, loss=self.loss)
+
+        return model
+
+    def _fit(self, X):
+        """Fit the classifier on the training set (X, y).
+
+        Parameters
+        ----------
+        X : np.ndarray of shape = (n_instances (n), n_dimensions (d), series_length (m))
```
Comment (TonyBagnall) [inline aeon/clustering/deep_learning/ae_fcn.py:237]: n_channels and n_timepoints

File: aeon/clustering/deep_learning/base.py [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+
+
+class BaseDeepClusterer(BaseClusterer, ABC):
+    """Abstract base class for deep learning time series clusterers.
+
+    Parameters
+    ----------
+    n_clusters: int, default=None
+        Number of clusters for the deep learnign model.
+    clustering_algorithm: str, default="kmeans"
```
Comment (TonyBagnall) [inline aeon/clustering/deep_learning/base.py:18]: spacing wrong

File: aeon/networks/ae_fcn.py [hunk 8]
```diff
@@ -0,0 +1,230 @@
+# -*- coding: utf-8 -*-
+"""Auto-Encoder using Fully Connected Neural Network (FCN)."""
+
+__author__ = ["James-Large", "AurumnPegasus", "hadifawaz1999"]
```
Comment (TonyBagnall) [inline aeon/networks/ae_fcn.py:4]: I think you are sole author of this

Comment (TonyBagnall) [inline aeon/networks/ae_fcn.py:4]: I think you are sole author of this

File: aeon/clustering/deep_learning/tests/test_deep_clusterer_base.py [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+import time
+
+import numpy as np
+import pytest
+
+from aeon.clustering.deep_learning.base import BaseDeepClusterer
+from aeon.utils._testing.collection import make_2d_test_data
+from aeon.utils.validation._dependencies import _check_soft_dependencies
+
+__author__ = ["achieveordie", "hadifawaz1999"]
```
Comment (TonyBagnall) [inline aeon/clustering/deep_learning/tests/test_deep_clusterer_base.py:14]: again, I think you are sole author

File: aeon/clustering/deep_learning/tests/test_deep_clusterer_base.py [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+
+    # test save last model to file than delete it
+
+    dummy_deep_clr.save_last_model_to_file()
+
+    os.remove("./" + last_file_name + ".hdf5")
+
+    # test summary of model
+
+    assert dummy_deep_clr.summary() is not None
```
Comment (TonyBagnall) [inline aeon/clustering/deep_learning/tests/test_deep_clusterer_base.py:102]: do you test _predict and _predict_proba anywhere?

File: aeon/clustering/deep_learning/ae_fcn.py [hunk 11]
```diff
@@ -0,0 +1,345 @@
...
+        except FileNotFoundError:
+            self.model_ = deepcopy(self.training_model_)
+
+        self._fit_clustering(X=X)
+
+        gc.collect()
+
+        return self
+
+    def _score(self, X, y=None):
```
Comment (TonyBagnall) [inline aeon/clustering/deep_learning/ae_fcn.py:305]: do you test all this?

**Review by TonyBagnall — CHANGES_REQUESTED** [review]
structure all looks good, comments about docstrings and testing, could you aim for 100% coverage?

File: aeon/clustering/deep_learning/ae_fcn.py (hunk 1 above)
Comment (hadifawaz1999) [reply aeon/clustering/deep_learning/ae_fcn.py]: fixed the docs issue everywhere

File: aeon/clustering/deep_learning/ae_fcn.py (hunk 2 above)
Comment (hadifawaz1999) [reply aeon/clustering/deep_learning/ae_fcn.py]: fixed

File: aeon/clustering/deep_learning/ae_fcn.py (hunk 3 above)
Comment (hadifawaz1999) [reply aeon/clustering/deep_learning/ae_fcn.py]: fixed

File: aeon/clustering/deep_learning/ae_fcn.py (hunk 4 above)
Comment (hadifawaz1999) [reply aeon/clustering/deep_learning/ae_fcn.py]: fixed this everywhere

File: aeon/clustering/deep_learning/ae_fcn.py (hunk 5 above)
Comment (hadifawaz1999) [reply aeon/clustering/deep_learning/ae_fcn.py]: fixed

File: aeon/clustering/deep_learning/ae_fcn.py (hunk 6 above)
Comment (hadifawaz1999) [reply aeon/clustering/deep_learning/ae_fcn.py]: fixed

File: aeon/clustering/deep_learning/base.py (hunk 7 above)
Comment (hadifawaz1999) [reply aeon/clustering/deep_learning/base.py]: fixed

File: aeon/networks/ae_fcn.py (hunk 8 above)
Comment (hadifawaz1999) [reply aeon/networks/ae_fcn.py]: fixed this, this is because of my copy paste of all classification code :)

Comment (hadifawaz1999) [reply aeon/networks/ae_fcn.py]: fixed

File: aeon/clustering/deep_learning/tests/test_deep_clusterer_base.py (hunk 9 above)
Comment (hadifawaz1999) [reply aeon/clustering/deep_learning/tests/test_deep_clusterer_base.py]: fixed

File: aeon/clustering/deep_learning/tests/test_deep_clusterer_base.py (hunk 10 above)
Comment (hadifawaz1999) [reply aeon/clustering/deep_learning/tests/test_deep_clusterer_base.py:101]: i test now these 2 functions

File: aeon/clustering/deep_learning/ae_fcn.py (hunk 11 above)
Comment (hadifawaz1999) [reply aeon/clustering/deep_learning/ae_fcn.py:304]: i test it now

**Comment by hadifawaz1999** [comment]
> structure all looks good, comments about docstrings and testing, could you aim for 100% coverage?

Thanks @TonyBagnall for all the comments, let me know if its okay now, coverage should be 100% now if i didn't miss anything

**Comment by TonyBagnall** [comment]
> > structure all looks good, comments about docstrings and testing, could you aim for 100% coverage?
> 
> Thanks @TonyBagnall for all the comments, let me know if its okay now, coverage should be 100% now if i didn't miss anything

could you resolve any comments that have been resolved please? ta

**Review by TonyBagnall — APPROVED** [review]
